text/0000-macro-at-most-once-rep.md

```diff
@@ -0,0 +1,171 @@
+
+```rust
+macro_rules! foo {
+  (do $b:block) => {
+    $b
+  }
+  (do $b1:block and $b2:block) => {
+    foo!($b1)
+    $b2
+  }
+}
+```
+
+Under this RFC, one would simply write:
+
+```rust
+macro_rules! foo {
+  (do $b1:block $(and $b2:block)?) => {
+    $b1
+    $($b2)?
+  }
+}
+```
+
+## Trailing commas
+
+Currently, the best way to make a rule tolerate trailing commas is to create another identical rule that has a comma at the end:
+
+```rust
+macro_rules! foo {
+  ($(pat),*,) => { foo!( $(pat),* ) };
+  ($(pat),*) => {
+    // do stuff
+  }
+}
+```
+
+or to allow multiple trailing commas:
+
+```rust
+macro_rules! foo {
+  ($(pat),* $(,)*) => {
+    // do stuff
+  }
+}
+```
+
+This is unergonomic and clutters up macro definitions needlessly. Under this RFC, one would simply write:
+
+```rust
+macro_rules! foo {
+  ($(pat),* $(,)?) => {
+    // do stuff
+  }
+}
+```
+
+Guide-level explanation
+-----------------------
+
+In Rust macros, you specify some "rules" which define how the macro is used and what it transforms to. For each rule, there is a pattern and a body:
+
+```rust
+macro_rules! foo {
+  (pattern) => { body }
+}
+```
+
+The pattern portion is composed of zero or more subpatterns concatenated together. One possible subpattern is to repeat another subpattern some number of times. This is useful when writing variadic macros (e.g. `println`):
+
+```rust
+macro_rules! println {
+  // Takes a variable number of arguments after the template
+  ($tempate:expr, $($args:expr),*) => { ... }
+}
+```
+which can be invoked like so:
+```rust
+println!("")           // 0 args
+println!("", foo)      // 1 args
+println!("", foo, bar) // 2 args
+...
+```
+
+The `*` in the pattern of this example indicates "0 or more repetitions". One can also use `+` for "at _least_ one repetition" or `?` for "at _most_ one repetition".
+
+In the body of a rule, one can specify to repeat some code for every occurence of the pattern in the invokation:
+
+```rust
+macro_rules! foo {
+  ($($pat:expr),*) => {
+    $(
+      println!("{}", $pat)
+    )* // Repeat for each `expr` passed to the macro
+  }
+}
+```
+
+The same can be done for `+` and `?`.
+
+The `?` operator is particularly useful for making macro rules with optional components in the invocation or for making macros tolerate trailing commas.
+
+Reference-level explanation
+---------------------------
+
+`?` is identical to `+` and `*` in use except that it represents "at most once" repetition.
+
+Drawbacks
+---------
+While there are grammar ambiguities, they can be easily fixed, as noted by @kennytm [here](https://internals.rust-lang.org/t/pre-rfc-at-most-one-repetition-macro-patterns/6557/2?u=mark-i-m):
+
+  > There is ambiguity: $($x:ident)?+ today matches a?b?c and not a+. Fortunately this is easy to resolve: you just look one more token ahead and always treat ?* and ?+ to mean separate by the question mark token.
```

Comment:
This should better be put back to reference-level explanation. You cannot implement `?` without clarifying this bit.